gemrb/tests/core/Strings/Test_String.cpp

```diff
@@ -23,6 +23,114 @@
 
 namespace GemRB {
 
+TEST(String_Test, StringFromView) {
+	String input{u"abc"};
+	auto result = StringFromView<String>(StringViewT<String>{input});
+	EXPECT_EQ(result, input);
+}
+
+TEST(String_Test, FindFirstOf) {
+	auto result = FindFirstOf(String{}, StringViewT<String>{});
+	EXPECT_EQ(result, String::npos);
+
+	result = FindFirstOf(String{u"abc"}, StringViewT<String>{}, 7);
+	EXPECT_EQ(result, String::npos);
+
+	result = FindFirstOf(String{u"abc"}, StringViewT<String>{u"cd"});
+	EXPECT_EQ(result, 2);
+
+	result = FindFirstOf(String{u"abcd"}, StringViewT<String>{u"c"}, 3);
+	EXPECT_EQ(result, String::npos);
+
+	result = FindFirstOf(String{u"abc"}, StringViewT<String>{u"de"});
+	EXPECT_EQ(result, String::npos);
+}
+
+TEST(String_Test, FindNotOf) {
+	String unit{u"abc"};
+
+	auto result = FindNotOf<String>(unit.cbegin(), unit.cend(), StringViewT<String>{u"def"});
+	EXPECT_EQ(result, unit.cbegin());
+
+	result = FindNotOf<String>(unit.cbegin(), unit.cend(), StringViewT<String>{u"abc"});
+	EXPECT_EQ(result, unit.cend());
+
+	result = FindNotOf<String>(unit.cbegin(), unit.cend(), StringViewT<String>{u"acd"});
+	EXPECT_EQ(result, unit.cbegin() + 1);
+
+	result = FindNotOf<String>(unit.cbegin(), unit.cend(), StringViewT<String>{});
+	EXPECT_EQ(result, unit.cbegin());
+}
+
+TEST(String_Test, FindFirstNotOf) {
+	auto result = FindFirstNotOf(String{}, StringViewT<String>{});
+	EXPECT_EQ(result, String::npos);
+
+	result = FindFirstNotOf(String{}, StringViewT<String>{}, String::npos);
+	EXPECT_EQ(result, String::npos);
```

Comment:
Writing one of these test cases, I was wondering whether the second check here makes any sense: https://github.com/gemrb/gemrb/blob/82e2b53b5fb6b0e60d2fe4dad964c609a0b3922c/gemrb/core/Strings/String.h#L75-L82

Reply:
it looks like it would yield a different return if it wasn't there: `return iter == s.end() ? STR::npos`. don't know if that matters.

Reply:
id check what the std lib does for its `std::string::find_first_not_of` and follow that.

Reply:
It says: _Position of the found character or [std::string::npos](https://en.cppreference.com/w/cpp/string/basic_string/npos) if no such character is found._

But I was looking at it logically, to check whether this was supposed to mean `sv` maybe: Since `pos` is non-negative, so at a length of 0 of `s`, the first check has already taken care of it.

Reply:
then we should probably also return `npos`

> the first check has already taken care of it.

For our known cases, probably. its template and `STR::size_type` doesn't have to be unsigned.